### PR TITLE
chore(apimachinery): deprecate meta.UnsafeGuessKindToResource

### PR DESCRIPTION
**Pull request by pmalek**
#### What type of PR is this?

/kind cleanup
/kind deprecation

#### What this PR does / why we need it:

This PR marks apimachinery's [`meta.UnsafeGuessKindToResource`](https://github.com/kubernetes/kubernetes/blob/6e0cb243d57592c917fe449dde20b0e246bc66be/staging/src/k8s.io/apimachinery/pkg/api/meta/restmapper.go#L123-L126) as deprecated given its unreliability and heuristic nature.

#### Which issue(s) this PR fixes:

Related to a discussion in https://github.com/kubernetes/kubernetes/pull/110053

#### Special notes for your reviewer:

Should I align all `k8s.io/gengo` versions to be the same as before this PR and not bump it or is bumping it from `k8s.io/gengo v0.0.0-20220902162205-c0856e24416d` to `k8s.io/gengo v0.0.0-20230306165830-ab3349d207d4` OK? I gather that I'd need to run `go mod vendor` then as well because running tests in `pkg/kubelet` complains about this.

#### Does this PR introduce a user-facing change?

```release-note
Mark apimachinery's `meta.UnsafeGuessKindToResource`. Users are suggested to use `k8s.io/gengo/namer`'s `NewAllLowercasePluralNamer()` instead.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/apimachinery/pkg/api/meta/restmapper.go [hunk 1]
```diff
@@ -62,6 +64,8 @@ type DefaultRESTMapper struct {
 	kindToScope          map[schema.GroupVersionKind]RESTScope
 	singularToPlural     map[schema.GroupVersionResource]schema.GroupVersionResource
 	pluralToSingular     map[schema.GroupVersionResource]schema.GroupVersionResource
+
+	namer namer.Namer
```
Comment (shivanshu1333) [inline staging/src/k8s.io/apimachinery/pkg/api/meta/restmapper.go:68]: may give fmt error

Comment (pmalek) [reply staging/src/k8s.io/apimachinery/pkg/api/meta/restmapper.go:68]: ACK. What's the fmt tool to check this and how can I trigger it? `gofmt` doesn't report anything

**Comment by pmalek** [comment]
I've checked locally an all UTs pass but there are some build related errors that I'm not sure what they are about. Can someone guide me how to approach this? https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/119924/pull-kubernetes-unit/1691019004237320192

File: staging/src/k8s.io/apimachinery/pkg/api/meta/restmapper.go [hunk 2]
```diff
@@ -24,6 +24,8 @@ import (
 
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/gengo/namer"
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/api/meta/restmapper.go:27]: gengo is for code-generation prior to build, not something we want to rely on at runtime

File: staging/src/k8s.io/client-go/testing/fixture_test.go [hunk 3]
```diff
@@ -278,12 +280,17 @@ func TestPatchWithMissingObject(t *testing.T) {
 func TestGetWithExactMatch(t *testing.T) {
 	scheme := runtime.NewScheme()
 	codecs := serializer.NewCodecFactory(scheme)
+	namer := namer.NewAllLowercasePluralNamer(nil)
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/testing/fixture_test.go:283]: using a namer without configurable exceptions just bakes in the special cases at another layer... same comment applies everywhere we're using `NewAllLowercasePluralNamer(nil)`

File: staging/src/k8s.io/apimachinery/pkg/api/meta/restmapper.go (hunk 2 above)
Comment (pmalek) [reply staging/src/k8s.io/apimachinery/pkg/api/meta/restmapper.go:27]: I see. What suggestion do you have on this one? `namer` ability to provide plural names was suggested in https://github.com/kubernetes/kubernetes/pull/110053#issuecomment-1585644795 as a replacement for `UnsafeGuessKindToResource`. What other options do we have?

File: staging/src/k8s.io/apimachinery/pkg/api/meta/restmapper.go [hunk 4]
```diff
@@ -93,11 +97,15 @@ func NewDefaultRESTMapper(defaultGroupVersions []schema.GroupVersion) *DefaultRE
 		defaultGroupVersions: defaultGroupVersions,
 		singularToPlural:     singularToPlural,
 		pluralToSingular:     pluralToSingular,
+		namer:                namer.NewAllLowercasePluralNamer(nil),
```
[the remark this answers is not in the document]
Comment (pmalek) [reply staging/src/k8s.io/apimachinery/pkg/api/meta/restmapper.go:100]: Pending response on the alternative for `namer` one question pops in my head: what improvement would you possibly see here? Make `NewDefaultRESTMapper` a variadic function accepting exceptions or add second "constructor" which allows to specify the exceptions keeping this intact (as is now)?

File: staging/src/k8s.io/apimachinery/pkg/api/meta/restmapper_test.go [hunk 5]
```diff
@@ -448,18 +446,33 @@ func TestKindToResource(t *testing.T) {
...
 		{Kind: "miss", Plural: "misses", Singular: "miss"},
 		// Add "s" otherwise
 		{Kind: "lowercase", Plural: "lowercases", Singular: "lowercase"},
+		// Take into account Endpoints exception.
+		{Kind: "Endpoints", Plural: "endpoints", Singular: "endpoints"},
 	}
-	for i, testCase := range testCases {
-		version := schema.GroupVersion{}
 
-		plural, singular := UnsafeGuessKindToResource(version.WithKind(testCase.Kind))
```
[the remark this answers is not in the document]
Comment (pmalek) [reply staging/src/k8s.io/apimachinery/pkg/api/meta/restmapper_test.go:459]: Done, reverted this but I retained the change using the subtests and adding the `Endpoints` test case.

File: staging/src/k8s.io/client-go/testing/fixture.go [hunk 6]
```diff
@@ -228,6 +231,9 @@ func NewObjectTracker(scheme ObjectScheme, decoder runtime.Decoder) ObjectTracke
 		decoder:  decoder,
 		objects:  make(map[schema.GroupVersionResource]map[types.NamespacedName]runtime.Object),
 		watchers: make(map[schema.GroupVersionResource]map[string][]*watch.RaceFreeFakeWatcher),
+		namer: namer.NewAllLowercasePluralNamer(map[string]string{
+			"Endpoints": "endpoints",
```
[the remark this answers is not in the document]
Comment (pmalek) [reply staging/src/k8s.io/client-go/testing/fixture.go:235]: Got it. I'd assume that you mean all the client "constructors" that utilize `NewObjectTracker` e.g. https://github.com/kubernetes/kubernetes/blob/88c8bcbb4aac388510adb1c7b1ed2b8a2f716996/staging/src/k8s.io/client-go/kubernetes/fake/clientset_generated.go#L136-L137.

What would you suggest we change/add to support said configurability? We can't introduce a breaking change (even if we wanted `NewSimpleClientset` already has a variadic parameter set `objects`). Do we want to add something like `NewSimpleClientsetWithExpections`? 🤔

[the remark this answers is not in the document]
Comment (liggitt) [reply staging/src/k8s.io/client-go/testing/fixture.go:235]:
```
// NewSimpleClientset returns a clientset that will respond with the provided objects.
// It's backed by a very simple object tracker that processes creates, updates and deletions as-is,
// without applying any validations and/or defaults. It shouldn't be considered a replacement
// for a real clientset and is mostly useful in simple unit tests.
// Equivalent to NewSimpleClientsetWithOptions(testing.Options{}, objects...)
func NewSimpleClientset(objects ...runtime.Object) *Clientset {
	return NewSimpleClientsetWithOptions(testing.Options{}, objects...)
}

// NewSimpleClientset returns a clientset that will respond with the provided objects.
// It's backed by a very simple object tracker that processes creates, updates and deletions as-is,
// without applying any validations and/or defaults. It shouldn't be considered a replacement
// for a real clientset and is mostly useful in simple unit tests.
func NewSimpleClientsetWithOptions(options testing.Options, objects ...runtime.Object) *Clientset {
	o := testing.NewObjectTrackerWithOptions(options, scheme, codecs.UniversalDecoder())
```

and 

```diff
diff --git a/staging/src/k8s.io/client-go/testing/fixture.go b/staging/src/k8s.io/client-go/testing/fixture.go
index 396840670fd..38ecf9745a6 100644
--- a/staging/src/k8s.io/client-go/testing/fixture.go
+++ b/staging/src/k8s.io/client-go/testing/fixture.go
@@ -216,18 +216,34 @@ type tracker struct {
 	// watchers' channel. Note that too many unhandled events (currently 100,
 	// see apimachinery/pkg/watch.DefaultChanSize) will cause a panic.
 	watchers map[schema.GroupVersionResource]map[string][]*watch.RaceFreeFakeWatcher
+
+	// ...
+	kindToResource map[schema.GroupVersionKind]schema.GroupVersionResource
 }
 
 var _ ObjectTracker = &tracker{}
 
+// ...
+type Options struct {
+	// KindToResource allows customizing the mapping for a given kind.
+	// If unset, UnsafeGuessKindToResource is used to attempt to derive the resource for a given kind.
+	KindToResource map[schema.GroupVersionKind]schema.GroupVersionResource
+}
+
 // NewObjectTracker returns an ObjectTracker that can be used to keep track
 // of objects for the fake clientset. Mostly useful for unit tests.
+// Equivalent to NewObjectTrackerWithOptions(Options{}, scheme, decoder)
 func NewObjectTracker(scheme ObjectScheme, decoder runtime.Decoder) ObjectTracker {
+	return NewObjectTrackerWithOptions(Options{}, scheme, decoder)
+}
+
+func NewObjectTrackerWithOptions(options Options, scheme ObjectScheme, decoder runtime.Decoder) ObjectTracker {
 	return &tracker{
 		scheme:   scheme,
 		decoder:  decoder,
 		objects:  make(map[schema.GroupVersionResource]map[types.NamespacedName]runtime.Object),
 		watchers: make(map[schema.GroupVersionResource]map[string][]*watch.RaceFreeFakeWatcher),
+
+		kindToResource: options.KindToResource,
 	}
 }
 
@@ -333,6 +349,8 @@ func (t *tracker) Add(obj runtime.Object) error {
 		return fmt.Errorf("no registered kinds for %v", obj)
 	}
 	for _, gvk := range gvks {
+               gvr, ok := t.kindToResource[gvk]
+               if !ok {
                        // NOTE: UnsafeGuessKindToResource is a heuristic and default match. The
                        // actual registration in apiserver can specify arbitrary route for a
                        // gvk. If a test uses such objects, it cannot preset the tracker with
                        // objects via Add(). Instead, it should trigger the Create() function
                        // of the tracker, where an arbitrary gvr can be specified.
-               gvr, _ := meta.UnsafeGuessKindToResource(gvk)
+                       gvr, _ = meta.UnsafeGuessKindToResource(gvk)
                        // Resource doesn't have the concept of "__internal" version, just set it to "".
                        if gvr.Version == runtime.APIVersionInternal {
                                gvr.Version = ""
                        }
+               }
 
 		err := t.add(gvr, obj, objMeta.GetNamespace(), false)
 		if err != nil {
```

**Comment by pohly** [comment]
/cc

Needed for https://github.com/kubernetes/kubernetes/pull/123938

**Comment by pohly** [comment]
I'm using a workaround in https://github.com/kubernetes/kubernetes/pull/123938 because it should better get merged before test freeze next week.

**Pull request closed without merging**